### PR TITLE
Flux GA release spec and long term support pledge

### PR DESCRIPTION
**Pull request by stefanprodan**
This PR introduces the release specifications for the Flux GA distribution.

The release documentation tries to answer questions like:
- How we release and version Flux, its CLI, controllers, Go packages and APIs
- How often are we going to release new Flux versions post-GA
- For which Flux versions we offer support and for how long
- Which Kubernetes versions are supported by a Flux release

Preview:
- [Flux distribution](https://github.com/fluxcd/flux2/blob/docs-release-spec/docs/release/flux.md)
- [Flux APIs and controllers](https://github.com/fluxcd/flux2/blob/docs-release-spec/docs/release/controllers.md)
- [Flux shared libraries](https://github.com/fluxcd/flux2/blob/docs-release-spec/docs/release/packages.md)
- [Flux release procedure](https://github.com/fluxcd/flux2/blob/docs-release-spec/docs/release/procedure.md)

### CODE REVIEW

File: docs/release/controllers.md
Comment (pjbgf) [inline docs/release/controllers.md:91]: Maybe?
```suggestion
controller release is included in a Flux CLI patch release.
```

File: docs/release/controllers.md
Comment (pjbgf) [inline docs/release/controllers.md:115]: We probably highlight that we follow Kubernetes' support policy, which at the moment is N-2 minor versions.

File: docs/release/controllers.md
Comment (bigkevmcd) [inline docs/release/controllers.md:124]: We could add some examples here similar to the Kubernetes docs ? https://kubernetes.io/releases/

File: docs/release/controllers.md
Comment (pjbgf) [inline docs/release/controllers.md:125]: Security patching and back-porting to the previous 2 minor releases, across all controllers, may be quite a big undertake given the size of the security team when considering all the additional work required around CVE handling and CVD.

From an OSS perspective, I think back porting to a single minor release sounds more achievable. Companies providing enterprise support for Flux users could bridge that gap for users requiring longer back-porting support.

File: docs/release/controllers.md
Comment (stefanprodan) [reply docs/release/controllers.md:91]: It's a Flux release which has nothing to do with the CLI binary. People can setup Flux to update itself from `ghcr.io/fluxcd/flux-manifests`, or they can use Flux TF provider or even the community maintained Helm charts.

File: docs/release/controllers.md
Comment (stefanprodan) [reply docs/release/controllers.md:125]: > Security patching and back-porting to the previous 2 minor releases, across all controllers, may be quite a big undertake given the size of the security team 

I don't think the actual back porting is a task that only the security team members can do. A security team member can create an issue asking the maintainers of a controller to do the back port and release. Now it's possible that maintainers don't have spare time to perform this task and no other contributor volunteers to do the back port. If this is the case, then back porting is not feasible due to a shortage of contributors, which is covered by the "depending on severity and feasibility".  I could expend on this if you want and explicitly call out "best-effort / community-driven".

File: docs/release/controllers.md
Comment (stefanprodan) [reply docs/release/controllers.md:53]: Kubernetes `v1` Service, StatefulSet even Pod got new fields without a version bump. Feature complete means that what's in the current RFC was implemented, it doesn't mean freezed for ever.

File: docs/release/controllers.md
Comment (pjbgf) [reply docs/release/controllers.md:125]: My main concern is about the security assurances we provide to our users as part of our CVE handling process.
During the embargo window the security team keep the CVE information private and patch all the supported versions - which currently is only one. It is only after the embargo is lifted and the CVE is public, that it could be back-ported on a "best-effort / community-driven" way. But by that time, users using older versions would already be exposed to an unpatched public CVE. 

Now for both security and bug fixes, the more we expand on this promise, the lesser our long-term capacity. 
I think we should pledge a more sustainable version range, and then mention that back-porting to older versions would be:
- provided by the community in a best-effort basis.
- provided by companies focused on Flux products and Services directed at enterprises (as in [this](https://fluxcd.io/support/#my-employer-needs-additional-help)).

But it would be good to hear from other security team members (cc @hiddeco @scottrigby).

File: docs/release/controllers.md
Comment (makkes) [inline docs/release/controllers.md:158]: I feel this would be a better fit for the website and then we can link from here. That way we can keep it current and have it a little more prominent.

File: docs/release/controllers.md
Comment (makkes) [inline docs/release/controllers.md:162]: It's good to have this documented centrally here. We should make sure to link here from the DEVELOPMENT.md files in the controllers' repos.

**Review by makkes — COMMENTED** [review]
Thanks @stefanprodan for putting all this info together. Great effort!

File: docs/release/controllers.md
Comment (stefanprodan) [reply docs/release/controllers.md:114]: We do https://github.com/fluxcd/flux2/blob/main/.github/workflows/e2e-arm64.yaml#L20

File: docs/release/controllers.md
Comment (stefanprodan) [reply docs/release/controllers.md:117]: Please suggest changes. What I wanted to say is that we'll do at least 3 releases (following Kubernetes), but besides that, we could do a minor release per week if we have new features to release.

File: docs/release/controllers.md
Comment (hiddeco) [inline docs/release/controllers.md:113]:
```suggestion
Flux controllers are at least released at the same rate as Kubernetes, following their cadence of three minor releases per year.
```

File: docs/release/controllers.md
Comment (stefanprodan) [reply docs/release/controllers.md:53]: Flux does not provides conversion webhooks, we avoid them by having the APIs backwards compatible.

File: docs/release/controllers.md
Comment (stefanprodan) [reply docs/release/controllers.md:119]: Please see the latest edits

File: docs/release/controllers.md
Comment (relu) [inline docs/release/controllers.md:127]:
```suggestion
Security fixes may be back-ported to those three minor versions as patch releases,
depending on severity and feasibility.
```

File: docs/release/controllers.md
Comment (relu) [inline docs/release/controllers.md:179]: It might be useful to add some information about handling backporting changes e.g. branch naming conventions, cherry-picking commits. Not sure where the right place for that would be.

File: docs/release/flux.md
Comment (relu) [inline docs/release/flux.md:64]:
```suggestion
Flux is at least released at the same rate as Kubernetes, following their cadence of three
minor releases per year. After each Kubernetes minor release, the CLI and all controllers are
tested against the latest Kubernetes version and are released approximately two weeks after Kubernetes.
The newly released Flux version offers support for Kubernetes N-2 minor versions.

Flux may have more than three minor releases per year if maintainers decide to ship a 
new feature or optimization ahead of schedule.
```

File: docs/release/flux.md
Comment (relu) [inline docs/release/flux.md:25]:
```suggestion
Release candidates are intended for testing new features or improvements before a final release.

In most cases, a maintainer will publish a release candidate for Flux users to test on their
staging clusters. Release candidates are not meant to be deployed in production unless advised
to do so by a maintainer.

Release candidates can be unstable and they are deprecated by subsequent RC or stable versions.
```

File: docs/release/flux.md
Comment (relu) [inline docs/release/flux.md:47]:
```suggestion
Minor releases are intended for backward-compatible feature additions and improvements.
Note that breaking changes may occur if required by a security vulnerability fix.

Minor releases are used when updating the Flux controllers or Kubernetes dependencies
from one minor version to another.

In effect, this means a Flux minor version will be released at least every four months after each
Kubernetes minor version release. To properly validate the Flux CLI and controllers against
the latest Kubernetes version, we reserve a time window of at least two weeks for end-to-end testing. 
```

File: docs/release/controllers.md
Comment (relu) [inline docs/release/controllers.md:118]:
```suggestion
Flux controllers are at least released at the same rate as Kubernetes, following their cadence of three
minor releases per year. After each Kubernetes minor release, all controllers are tested against the latest
Kubernetes version and then released approximately two weeks after Kubernetes.
The newly released controllers offer support for Kubernetes N-2 minor versions.

```

File: docs/release/controllers.md
Comment (relu) [inline docs/release/controllers.md:40]: It's not clear where it's safe to be used considering the above paragraph regarding alpha versions only being used in testing environments.
```suggestion
A beta version API e.g. `v2beta1` is considered well-tested and safe to use in production.
```

File: docs/release/flux.md
Comment (hiddeco) [inline docs/release/flux.md:15]:
```suggestion
- `vX.Y.Z-RC.W` release candidates e.g. `v2.0.0-rc.1`
- `vX.Y.Z` stable releases e.g. `v2.0.0`

The Flux project maintains release branches for the most recent three minor releases
e.g. `release/2.0.x`, `release/2.1.x` and `release/2.2.x`.
```

- We have tagged our RC with `-rc.X`, so lets use that as a proper example.
- I prefer `/` "namespacing" on repositories where features are also developed in branches, and `-` is used in those names. This also eases things like branch protection and automation without creating false positives.

File: docs/release/controllers.md
Comment (hiddeco) [inline docs/release/controllers.md:103]:
```suggestion
In effect, this means a new minor version will at least be released for all Flux controllers approximately every four months, following each Kubernetes minor version release. To properly validate the controllers against the latest Kubernetes version, we typically allocate a time window of at least two weeks for end-to-end testing of Flux controllers.

It is worth noting that in certain scenarios where project dependencies are not in sync with the Kubernetes version or conflicts arise, this two-week timeframe may prove insufficient, requiring additional time to address the issues appropriately.
```

File: docs/release/controllers.md
Comment (hiddeco) [inline docs/release/controllers.md:164]: I would move this to a separate document (e.g. `process.md`), as the audience is different.

File: docs/release/controllers.md
Comment (hiddeco) [reply docs/release/controllers.md]: Think there is a separation here between two things:

1. CVEs which are directly the result of (not so) happy mistakes in our code.
2. CVEs from upstream dependencies.

---

For scenario 1, having means to automatically back-port things through GitHub labeling may work in about 75% of the cases. I think we could pull this of in a relatively effortless way, as I have seen other projects do.

In terms of this exposing indicators around security issues, the same could be said for closely following commits and/or pull requests. In addition, given it is likely that once this is in place we also leverage this to fix critical (non-security) bugs. Any indicators would blend together with non-CVE things.

This makes me lean towards the idea of this being a non-issue in terms of exposure, or at least outweighing the discomfort of people having to closely follow our minor ranges. Especially given our (at times) aggressive release cadence.

---

For scenario 2, I think this is more of an issue in terms of human workload. Every minor range of a controller may have their own supported minor range of some dependencies, and some of these dependencies may or may not have their procedures around back-porting security fixes. This means that:

1. We have to manually update versions within the minor range of the dependency.
2.  We may not be able to resolve the issue, as it would require updating beyond the point of a version range of the dependency which is accepted within the range of the controller.
3. All of this would be an exponential effort.

---

All together, I think that we should pledge `n-2` for scenario 1, but keep scenario two at `n-1` + best effort.

File: docs/release/controllers.md
Comment (stefanprodan) [reply docs/release/controllers.md]: We can't, release dates are not set in stone, some dependency like client-go could break compatibility with other Flux deps such as Helm, so we need to wait for all deps to sync.

**Comment by stefanprodan** [comment]
On hold till we figure out the cherry-pick automation needed for the new release strategy.

File: docs/release/controllers.md
Comment (makkes) [inline docs/release/controllers.md:26]:
```suggestion
The schema of objects may change in incompatible ways in a later API version.
```

This aligns the wording with the one in the "beta version" section.

File: docs/release/controllers.md
Comment (makkes) [inline docs/release/controllers.md:30]:
```suggestion
A deprecated alpha version is subject to removal after a three month period.
```

File: docs/release/controllers.md
Comment (makkes) [inline docs/release/controllers.md:51]:
```suggestion
A stable API version, e.g. `v2`, is considered feature complete.
```

File: docs/release/controllers.md
Comment (makkes) [inline docs/release/controllers.md:57]:
```suggestion
A stable API version becomes deprecated once a subsequent stable version is released.
```

File: docs/release/controllers.md
Comment (makkes) [inline docs/release/controllers.md:58]:
```suggestion
Stable API versions are not subject to removal in any future release within a controller major version.
```

Makes it clearer that we're talking about a single major version and not about the possibility that a stable API version may be removed in a subsequent major controller version.

File: docs/release/controllers.md
Comment (makkes) [inline docs/release/controllers.md:68]:
```suggestion
- `vX.Y.Z-rc.W` release candidates e.g. `v1.0.0-rc.1`
```
This is how we called releases up until now.

File: docs/release/controllers.md
Comment (makkes) [inline docs/release/controllers.md:89]:
```suggestion
If a security fix requires a breaking change, then a minor release will provide the fix.
```
Makes it more explicit that we're talking about security fixes here and not about CVEs in particular. Except you wanted to pin this to CVEs?!

File: docs/release/controllers.md
Comment (makkes) [inline docs/release/controllers.md:122]: Not relevant to this PR getting merged or not but we have to make sure that we have a process in place that ensures we're following what is stated here. At the moment we have no formal release management or calendar or anything like that in place so as it is now it would depend on someone recalling to do this after each K8s release.

File: docs/release/flux.md
Comment (makkes) [inline docs/release/flux.md:11]:
```suggestion
- `vX.Y.Z-rc.W` release candidates e.g. `v2.0.0-rc.1`
```

File: docs/release/packages.md
Comment (makkes) [inline docs/release/packages.md:14]:
```suggestion
- `NAME/vX.Y.Z-rc.W` release candidates e.g. `runtime/v1.0.0-rc.1`
```

**Review by makkes — APPROVED** [review]

File: docs/release/procedure.md
Comment (aryan9600) [inline docs/release/procedure.md:260]:
```suggestion
#### Distribution of minor releases
```
or
```suggestion
#### Distributing minor releases
```
maybe? (same goes for the below sections)

File: docs/release/procedure.md
Comment (aryan9600) [inline docs/release/procedure.md:232]:
```suggestion
Preview releases are made by triggering the `release` GitHub Actions workflow on
```

File: docs/release/procedure.md
Comment (aryan9600) [inline docs/release/procedure.md:25]:
```suggestion
- [Backport changes for patch releases](#backport-changes-for-patch-releases)
  + [Manual backporting](#manual-backporting)
```

File: docs/release/procedure.md
Comment (stefanprodan) [inline docs/release/procedure.md:267]:
```suggestion
   `release/v2.2.x`, and push it to the remote repository.
```

Comment (stefanprodan) [reply docs/release/procedure.md:267]: Please let's use `v2` for the distribution examples here and below.

File: docs/release/procedure.md
Comment (hiddeco) [reply docs/release/procedure.md:260]: They're "minor release of the (Flux) distribution". Maybe something like the following would be more clear?

> Distribution: minor releases

> Controller: minor releases

> etc.

The main reason to work with a prefix is to have unique refs for the titles (which can then be used in the ToC).

File: docs/release/controllers.md
Comment (hiddeco) [inline docs/release/controllers.md:120]:
```suggestion
Flux controllers are _at least_ released at the same rate as Kubernetes, following their cadence of three
```

to emphasize this more, as I read over it (again).

File: docs/release/flux.md
Comment (hiddeco) [inline docs/release/flux.md:58]:
```suggestion
Flux is _at least_ released at the same rate as Kubernetes, following their cadence of three
```

same as above.

File: docs/release/flux.md
Comment (hiddeco) [inline docs/release/flux.md:47]:
```suggestion
In effect, this means a Flux minor version will be released at least every four months, after each
Kubernetes minor version release. For in-depth information about this, please refer to the
[release cadence](#release-cadence) section of this document.
```

File: docs/release/controllers.md
Comment (hiddeco) [inline docs/release/controllers.md:109]: I feel this should probably be moved to the release cadence section, and link to instead. Basically the same approach as I suggested in the `flux.md` document.

File: docs/release/procedure.md
Comment (aryan9600) [reply docs/release/procedure.md:266]: yeah "Distribution: minor releases" seems good

**Review by aryan9600 — APPROVED** [review]
thanks @stefanprodan!

**Review by hiddeco — APPROVED** [review]

**Pull request merged**
